agents/geminidataanalytics/colabs/geminidataanalytics_public_SDK_demo_|_vPublished.ipynb

```diff
@@ -0,0 +1,1538 @@
+        "\n",
+        "def cleanup_path(input_string):\n",
+        "  \"\"\"Removes bracketed numbers and preceding dots from an input string.\n",
+        "\n",
+        "  Args:\n",
+        "    input_string: The input string.\n",
+        "\n",
+        "  Returns:\n",
+        "    The string with bracketed numbers and preceding dots removed.\n",
+        "  \"\"\"\n",
+        "  cleaned_string = re.sub(r\"\\[\\d+]\", \"\", input_string)\n",
+        "  if cleaned_string.startswith(\".\"):\n",
+        "    cleaned_string = cleaned_string[1:]\n",
+        "  return cleaned_string\n",
+        "\n",
+        "def is_valid_yaml(yaml_string):\n",
+        "    try:\n",
+        "        yaml.safe_load(yaml_string)\n",
+        "        return True\n",
+        "    except ScannerError:\n",
+        "        return False\n",
+        "    except Exception as e:\n",
+        "        print(f\"An unexpected error occurred: {e}\")\n",
+        "        return False\n",
+        "\n",
+        "def print_counter_desc(data):\n",
+        "    # Count the occurrences of each item in the list.\n",
+        "    item_counts = Counter(data)\n",
+        "\n",
+        "    # Sort the items in descending order based on their counts.\n",
+        "    sorted_items = sorted(item_counts.items(), key=lambda item: item[1], reverse=True)\n",
+        "\n",
+        "    # Print the items and their counts.\n",
+        "    for item, count in sorted_items:\n",
+        "        print(f\"{item}: {count}\")\n",
+        "\n",
+        "def extract_insights_from_system_instruction(yaml_string):\n",
+        "\n",
+        "  \"\"\"\n",
+        "    Extracts insights into used fields from a YAML string.\n",
+        "\n",
+        "    Args:\n",
+        "      yaml_string: The YAML string to analyze.\n",
+        "  \"\"\"\n",
+        "\n",
+        "  if not is_valid_yaml(yaml_string):\n",
+        "      print(\"Received Invalid or not a YAML format. Cannot extract insights.\")\n",
+        "      return\n",
+        "\n",
+        "  try:\n",
+        "      data = yaml.safe_load(yaml_string)\n",
+        "      used_fields = []\n",
+        "      def traverse(node, path=\"\"):\n",
+        "            if isinstance(node, dict):\n",
+        "                for key, value in node.items():\n",
+        "                    new_path = f\"{path}.{key}\" if path else key\n",
+        "                    traverse(value, new_path)\n",
+        "            elif isinstance(node, list):\n",
+        "                for i, item in enumerate(node):\n",
+        "                    new_path = f\"{path}[{i}]\"\n",
+        "                    traverse(item, new_path)\n",
+        "            else:\n",
+        "                used_fields.append(cleanup_path(path))\n",
+        "\n",
+        "      traverse(data)\n",
+        "      print(\"Insights:\")\n",
+        "      print(\"Number of unique fields used: \", len(set(used_fields)))\n",
```

Comment:
Is there a specific reason that yaml is used for defining the system instructions? Is it required for the data science agent to work?